File: src/main/java/kome/hatsuneakiko/pastenbt/ServerboundPastePacket.java
package kome.hatsuneakiko.pastenbt;

import net.minecraft.core.BlockPos;
import net.minecraft.core.registries.Registries;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtIo;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
import net.minecraftforge.fml.loading.FMLPaths;
import net.minecraftforge.network.NetworkEvent;

import java.io.File;
import java.io.FileInputStream;
import java.util.List;
import java.util.function.Supplier;

public class ServerboundPastePacket {
    private final String name;
    private final Rotation rotation;
    private final SnapMode snapMode;

    public ServerboundPastePacket(String name, Rotation rotation, SnapMode snapMode) {
        this.name = name;
        this.rotation = rotation;
        this.snapMode = snapMode;
    }

    public static void encode(ServerboundPastePacket msg, FriendlyByteBuf buf) {
        buf.writeUtf(msg.name);
        buf.writeEnum(msg.rotation);
        buf.writeEnum(msg.snapMode);
    }

    public static ServerboundPastePacket decode(FriendlyByteBuf buf) {
        return new ServerboundPastePacket(buf.readUtf(), buf.readEnum(Rotation.class), buf.readEnum(SnapMode.class));
    }

    public static void handle(ServerboundPastePacket msg, Supplier<NetworkEvent.Context> ctxGetter) {
        NetworkEvent.Context ctx = ctxGetter.get();
        ServerPlayer player = ctx.getSender();
        if (player == null || !ModConfig.canUse(player)) {
            ctx.setPacketHandled(true);
            return;
        }

        new Thread(() -> {
            try {
                File nbtFile = FMLPaths.GAMEDIR.get().resolve("structures/" + msg.name + ".nbt").toFile();
                if (!nbtFile.exists()) {
                    player.sendSystemMessage(Component.literal("File not found: " + msg.name));
                    return;
                }
                try (FileInputStream fis = new FileInputStream(nbtFile)) {
                    CompoundTag nbt = NbtIo.readCompressed(fis);
                    StructureTemplate template = new StructureTemplate();
                    template.load(player.serverLevel().holderLookup(net.minecraft.core.registries.Registries.BLOCK), nbt);
                    BlockPos origin = msg.snapMode.snap(player.blockPosition());
                    StructurePlaceSettings settings = new StructurePlaceSettings().setRotation(msg.rotation);
                    List<StructureTemplate.StructureBlockInfo> blocks = template.filterBlocks(origin, settings, null);
                    ctx.enqueueWork(() -> {
                        PlacementManager.add(new PlacementManager.PlacementTask(player.getUUID(), player.serverLevel(), blocks));
                        player.sendSystemMessage(Component.literal("§b[Java 17 Async]§r Prepared: " + msg.name));
                    });
                }
            } catch (Exception e) {
                player.sendSystemMessage(Component.literal("Error: " + e.getMessage()));
            }
        }).start();
        ctx.setPacketHandled(true);
    }
}

File: src/main/java/kome/hatsuneakiko/pastenbt/ServerboundPastePacket.kt
package kome.hatsuneakiko.pastenbt

import net.minecraft.core.registries.Registries
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtIo
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.chat.Component
import net.minecraft.world.level.block.Rotation
import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate
import net.minecraftforge.fml.loading.FMLPaths
import net.minecraftforge.network.NetworkEvent
import java.io.FileInputStream
import java.util.function.Supplier

class ServerboundPastePacket(
    val name: String,
    val rotation: Rotation,
    val snapMode: SnapMode
) {
    companion object {
        @JvmStatic
        fun encode(msg: ServerboundPastePacket, buf: FriendlyByteBuf) {
            buf.writeUtf(msg.name)
            buf.writeEnum(msg.rotation)
            buf.writeEnum(msg.snapMode)
        }

        @JvmStatic
        fun decode(buf: FriendlyByteBuf): ServerboundPastePacket {
            return ServerboundPastePacket(buf.readUtf(), buf.readEnum(Rotation::class.java), buf.readEnum(SnapMode::class.java))
        }

        @JvmStatic
        fun handle(msg: ServerboundPastePacket, ctxGetter: Supplier<NetworkEvent.Context>) {
            val ctx = ctxGetter.get()
            val player = ctx.sender ?: return
            if (!ModConfig.canUse(player)) return

            Thread.ofVirtual().start {
                try {
                    val nbtFile = FMLPaths.CONFIGDIR.get().resolve("structures/${msg.name}.nbt").toFile()
                    if (!nbtFile.exists()) {
                        player.sendSystemMessage(Component.literal("File not found: ${msg.name}"))
                        return@start
                    }

                    FileInputStream(nbtFile).use { fis ->
                        val nbt = NbtIo.readCompressed(fis)
                        val template = StructureTemplate()
                        template.load(player.serverLevel().holderLookup(Registries.BLOCK), nbt)

                        val origin = msg.snapMode.snap(player.blockPosition())
                        val settings = StructurePlaceSettings().setRotation(msg.rotation)
                        val blocks = template.filterBlocks(origin, settings, null)

                        ctx.enqueueWork {
                            PlacementManager.add(PlacementManager.PlacementTask(player.uuid, player.serverLevel(), blocks))
                            player.sendSystemMessage(Component.literal("§b[Kotlin Coroutine]§r Prepared: ${msg.name}"))
                        }
                    }
                } catch (e: Exception) {
                    player.sendSystemMessage(Component.literal("Error: ${e.message}"))
                }
            }
            ctx.packetHandled = true
        }
    }
}
